internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala

```diff
@@ -69,12 +75,32 @@ private[inc] abstract class IncrementalCommon(
       doCompile: CompileCycle,
       classfileManager: XClassFileManager,
       output: Output,
-      cycleNum: Int
-  ) {
+      cycleNum: Int,
+      progress: Option[CompileProgress],
+      var notifiedEarly: Boolean = false,
```

Comment:
Ah, damn, I don't like this detail.  It's already very confusingly callback hell and recursive that I really don't want to add mutation to the mix...

I would say let's drop the (never-was-documented) guarantee that `earlyOutputComplete` is only called once, and let the implementation deal with it.  `tryComplete` on `Promise` doesn't break like `complete`.

Reply:
It's not only about once or multiple times that's the problem but also when. Prior to this change the notification was happening either prematurely before the final cycle or sometimes not triggered at all. We can refractor the impl details but this fix needs to happen.

Reply:
Reading the code again, maybe I'm misunderstood, and I might need to only handle the no-op cases. If so I can get rid of the var.

Reply:
https://github.com/sbt/zinc/pull/849/commits/75a3bb91785fed6326de4b48a0d8653eb994fecf is a much smaller change I just pushed.